libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1808 @@
+  using __size_type    = __inplace_vector_size_type<_Capacity>;
+  using reference      = _Tp&;
+  using iterator       = _Tp*;
+  using const_iterator = const _Tp*;
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr __inplace_vector_base() = default;
+  _LIBCUDACXX_INLINE_VISIBILITY
+  __inplace_vector_base(const __inplace_vector_base& __other) noexcept(_CCCL_TRAIT(is_nothrow_copy_constructible, _Tp))
+      : __inplace_vector_destruct_base<_Tp, _Capacity, _Spec>()
+  {
+    __uninitialized_copy(__other.begin(), __other.end(), begin());
+  }
+  _LIBCUDACXX_INLINE_VISIBILITY
+  __inplace_vector_base(__inplace_vector_base&& __other) noexcept(_CCCL_TRAIT(is_nothrow_move_constructible, _Tp))
+      : __inplace_vector_destruct_base<_Tp, _Capacity, _Spec>()
+  {
+    __uninitialized_move(__other.begin(), __other.end(), begin());
+    __other.__destroy(__other.begin(), __other.end());
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_base& operator=(const __inplace_vector_base& __other) noexcept(
+    _CCCL_TRAIT(is_nothrow_copy_constructible, _Tp) && _CCCL_TRAIT(is_nothrow_copy_assignable, _Tp))
+  {
+    if (__other.size() < size())
+    {
+      const auto __new_end = _CUDA_VSTD::copy(__other.begin(), __other.end(), begin());
+      __destroy(__new_end, end());
+    }
+    else
+    {
+      _CUDA_VSTD::copy(__other.begin(), __other.begin() + size(), begin());
+      __uninitialized_copy(__other.begin() + size(), __other.end(), end());
+    }
+    return *this;
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY __inplace_vector_base& operator=(__inplace_vector_base&& __other) noexcept(
+    _CCCL_TRAIT(is_nothrow_move_constructible, _Tp) && _CCCL_TRAIT(is_nothrow_move_assignable, _Tp))
+  {
+    if (__other.size() < size())
+    {
+      const auto __new_end = _CUDA_VSTD::move(__other.begin(), __other.end(), begin());
+      __destroy(__new_end, end());
+    }
+    else
+    {
+      _CUDA_VSTD::move(__other.begin(), __other.begin() + size(), begin());
+      __uninitialized_move(__other.begin() + size(), __other.end(), end());
+    }
+
+    // Need to reset the other container
+    __other.__destroy(__other.begin(), __other.end());
+
+    return *this;
+  }
+
+  // [containers.sequences.inplace.vector.members] size/capacity
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void __change_size(const ptrdiff_t __count) noexcept
+  {
+    this->__size_ += static_cast<__size_type>(__count);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr size_type size() const noexcept
+  {
+    return static_cast<size_type>(this->__size_);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr bool empty() const noexcept
+  {
+    return this->__size_ == 0;
+  }
+
+  // [containers.sequences.inplace.vector.data], data access
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr _Tp* data() noexcept
+  {
+    return _CUDA_VSTD::launder(reinterpret_cast<_Tp*>(this->__elems_));
+  }
```

Comment:
`constexpr` never works here (and in many other places) due to `reinterpret_cast`. If we want to use the storage-providing mechanism, perhaps `constexpr` should be dropped.

Reply:
Oh yeah, that is a incorrect copy error